openedx/core/djangoapps/xblock/README.rst

```diff
@@ -0,0 +1,94 @@
+* To provide a complete runtime environment for XBlocks, with little or no support for legacy XModules
+   - Avoiding XModule support as much as possible keeps the code cleaner
+   - All XModule compatibility code that we do include must be isolated within the ``shims.py`` file for eventual removal
+   - As a firm rule, any block with code split into separate ``Module`` and ``Descriptor`` classes will not be supported.
+* To support learning experiences other than courses
+   - The new runtime is built around the concept of "learning contexts," which are more generic and flexible than courses. See `Learning Contexts`_ below.
+   - Additional learning contexts and related functionality can be added via plugins.
+   - The new runtime makes as few assumptions as possible about how XBlocks are structured (doesn't expect a rooted tree hierarchy like every course and library in the old runtime must have).
+* To avoid any frontend code
+   - This new runtime is API-first, designed to be consumed by SPA frontends or mobile apps rather than rendering HTML into templates and sending them to the browser. Since it is based on XBlocks, there are some circumstances where it returns HTML generated by the XBlock views themselves, but it returns that HTML as an API response only, wrapped in a JSON "Fragment" object. In no other circumstances does this new runtime produce HTML.
+   - This is in contrast to the old runtime in which key parts of the LMS UI were added to HTML output by the runtime and/or the XBlocks themselves (e.g. the ``sequential`` XBlock rendered next/prev navigation buttons).
+* To provide first-class support for anonymous usage of XBlocks (use by unregistered users)
+   - If the user is unregistered, field data will be saved into the user's django session rather than the database.
+* To support sandboxed execution of XBlocks
+   - The runtime API provides a mechanism for calling XBlock handlers without session authentication (using a secure token in the URL), so that XBlocks can be run in sandboxed IFrames that cannot access the user's LMS cookies or interact with the LMS other than through the XBlock JavaScript API.
+
+Code Layout and Separation of Concerns
+--------------------------------------
+
+All of this code is found in this ``openedx.core.djangoapps.xblock`` folder:
+
+* `api.py <./api.py>`_ : **Python API** for this new XBlock runtime. Provides methods for loading and interacting with XBlocks given a user and a usage ID, such as: ``load_block()``, ``get_block_metadata()``, ``render_block_view()``, and ``get_handler_url()``
+
+  - Implementation differences in how Studio and the LMS implement these APIs are encapsulated in ``lms_impl.py`` and ``studio_impl.py``
+* `rest_api <./rest_api/>`_ : **REST API** for this new XBlock runtime. Provides `endpoints <./rest_api/urls.py>`_ for things like getting an XBlock's metadata, rendering an XBlock's view, calling an XBlock's handler, etc.
```

Comment:
Interesting separation. At first this was a little jarring to me because I'm used to Django app stuff being at the center of everything (i.e. at the top level of this package), but I get that the REST API is only a small part of the whole here, so this works for me.

Reply:
> Interesting separation.

Let's see how it goes and we can always slice and dice it differently if needed. I'm not married to this approach. I started with a different approach but gravitated to this one because:
(a) I always struggle to find where the runtime code is in the current codebase (`cms/djangoapps/contentstore/views/preview.py` and `lms/djangoapps/courseware/module_render.py` are they key files I have trouble finding, even after all these years) so I "feel" like I don't want to follow that same approach
(b) It makes it crystal clear where the studio/lms-specific code/parameters are, and it's really easy to compare one to the other